### ISSUE TITLE
Conflict between `pattern_type_mismatch` and  `needless_borrowed_reference` in for loop.

### ISSUE THREAD

**Issue by GilShoshan94** [rust-lang/rust-clippy#10376]
### Summary

Hi,

Catch 22 situation when fixing one lint trigger the other.

I provided a minimal code to reproduce the situation.

### Reproducer

I tried this code:

```rust
    #[warn(clippy::pattern_type_mismatch)]
    #[warn(clippy::needless_borrowed_reference)]
    let test1 = {
        let args: Vec<(i32, i32)> = vec![(1, 0), (2, 0), (3, 1)];

        let mut res = Vec::new();

        // Uncomment one of the 2 following lines.
        for (val, cat) in &args {
        // for &(ref val, ref cat) in &args {
            match *cat {
                0 => res.push("Category 0"),
                1 => res.push("Category 1"),
                _ => res.push("Unknown Category"),
            };
        }
        res
    };
```

I expected to see this happen:

I am not sure, I try to rely on clippy to learn the good idiomatic way to write Rust code.
I guess I expected the `for &(ref val, ref cat) in &args {` to be the better way and not have any lint. I am not sure.

Instead, this happened:
I am trapped in a catch 22:

* If I use `for (val, cat) in &args {`, I get the `pattern_type_mismatch` lint:
![image](https://user-images.githubusercontent.com/29143941/219956252-526a65a5-9912-451a-a182-6200b3f4c467.png)

* If I use `for &(ref val, ref cat) in &args {`, I get the `needless_borrowed_reference` lint: 
![image](https://user-images.githubusercontent.com/29143941/219956323-995a18e7-b7f2-44fd-9a19-aa137c899655.png)

### Version

```text
rustc 1.67.1 (d5a82bbd2 2023-02-07)
binary: rustc
commit-hash: d5a82bbd26e1ad8b7401f6a718a9c57c96905483
commit-date: 2023-02-07
host: x86_64-pc-windows-msvc
release: 1.67.1
```

**Comment by Jarcho** [comment]
`pattern_type_mismatch` is a restriction lint, and as such is not recommended for general use. Restriction lints are intended for either an unusual coding style (generally by restricting certain language features), or for narrow use cases that most code doesn't fall under.

Basically, this isn't a bug. Restriction lints are allowed to contradict other lints.

**Comment by GilShoshan94** [comment]
@Jarcho
Thank you for the reply, I get it that `needless_borrowed_reference` is a complexity lint, on warn by default, and `pattern_type_mismatch` is a restriction lint, on allow by default.

I also understand that some lints in the restriction lint contradict one each other because their are "chose one" style, such as `separated_literal_suffix` vs `unseparated_literal_suffix`.

I still find it odd to have a restriction lint to conflict with other "__default__" lints. It's surprising.

At least it should be mention in the README that restriction lints can conflict with the "__default__" ones.

This is the only mention about the restriction lints currently (that I could find easily):

> The [lint list](https://rust-lang.github.io/rust-clippy/master/index.html) also contains "restriction lints", which are for things which are usually not considered "bad", but may be useful to turn on in specific cases. These should be used very selectively, if at all.

**Comment by mickvangelderen** [comment]
Ran into this. I wanted to enable `pattern_type_mismatch` to ensure that I'm taking a reference where I expect to take a reference and to enforce this by explicitly writing `&(ref a, ref b, ..)`. Whether that's worth sacrificing readability, I'm also not sure. I do think it is unfortunate that this lint conflicts with `needless_borrowed_reference`.

**Comment by antoyo** [comment]
What's the suggestion here?
Should users of `pattern_type_mismatch` add the following:

```rust
#![allow(clippy::needless_borrowed_reference)]
```

If so, could that be done automatically by clippy?
